pkg/api/pod/util.go

```diff
@@ -23,6 +23,18 @@ import (
 	"k8s.io/kubernetes/pkg/util/intstr"
 )
 
+const (
+	// The annotation value is a string specifying the hostname to be used for the pod e.g 'my-webserver-1'
+	PodHostnameAnnotation = "pod.beta.kubernetes.io/hostname"
+
+	// The annotation value is a string specifying the subdomain e.g. "my-web-service"
+	// If specified, on the the pod itself, "<hostname>.my-web-service.<namespace>.svc.<cluster domain>" would resolve to
+	// the pod's IP.
+	// If there is a headless service named "my-web-service" in the same namespace as the pod, then,
+	// <hostname>.my-web-service.<namespace>.svc.<cluster domain>" would be resolved by the cluster DNS Server.
+	PodSubdomainAnnotation = "pod.beta.kubernetes.io/subdomain"
```

Comment:
> > +const (
> > + // The annotation value is a string specifying the hostname to be used for the pod e.g 'my-webserver-1'
> > + PodHostnameAnnotation = "pod.beta.kubernetes.io/hostname"
> > +
> > + // The annotation value is a string specifying the subdomain e.g. "my-web-service"
> > + // If specified, on the the pod itself, "<hostname>.my-web-service.<namespace>.svc.<cluster domain>" would resolve to
> > + // the pod's IP.
> > + // If there is a headless service named "my-web-service" in the same namespace as the pod, then,
> > + // <hostname>.my-web-service.<namespace>.svc.<cluster domain>" would be resolved by the cluster DNS Server.
> > + PodSubdomainAnnotation = "pod.beta.kubernetes.io/subdomain"
>
> @liggitt and I are still confused on this.
>
> Also, the pod must be an endpoint for the headless service...
>
> Why use a headless service? Why not a service that matches a single pod by label?

I think we started with headless because it seems like the most
obvious fit, but non-headless could be plausible.  Think of it this
way: A "regular" service means "all my backends are fungible, I don't
care which you use".  The DNS lookup returns just the cluster IP.  A
headless service means "my individual backends might have some meaning
that kube doesn't know about".  The DNS lookup returns the set of A
records for each backend.  In this PR we are adding the ability to
give meaningful names to those backends.  We could give meaningful
names to backends of a non=-headless Service, but that's not nearly as
obviously useful.

Secondly, we don't want to have DNS be watching Pods, if we can avoid
it.  So this feature actually comes in two parts.  The pod can request
a hostname and domainname, such that `hostname -f` DTRT. If and only
if you *also* have a (headless) service that matches domainname, will
you get a DNS record.  In essence, the pod domainname is how the pod
indicates that it has a "primary" service that governs it.   The most
obvious case here is a petset.

This feature should allow a petset controller to stamp out pods with
hostname and domainname set, create the headless service, and voila -
petset DNS works.

> This allows the pod to declare that it expects an FQDN and that the form of the FQDN will match the names that we add to DNS for service endpoints.
>
> This confuses us. Can you explain why this is needed? It's less obvious to me than the pod hostname.

It's a little tricky that the pod FQDN is ".svc.cluster.local".  This
is first and foremost because we don't want to make DNS watch pods.
DNS is fundamentally ABOUT services.  This lets a pod "opt-in" to
pod-DNS without encouraging every single pod to choose a unique
hostname and blow out DNS.